DataManagementSystem/private/FTS3Utilities.py

```diff
@@ -96,9 +99,50 @@ def generatePossibleTransfersBySources( ftsFiles, allowedSources = None ):
       if allowedSources and se not in allowedSources:
         continue
 
-      groupBySource.setdefault( se, [] ).append( ftsFile )
+      groupBySource.setdefault(se, []).append(ftsFile)
+
+  return S_OK(groupBySource)
+
+
+def selectUniqueRandomSource(ftsFiles, allowedSources=None):
+  """
+      For a list of FTS3files object, select a random source, and group the files by source.
+
+      :param allowedSources : list of allowed sources
+      :param ftsFiles : list of FTS3File object
+
+      :return  S_OK({ sourceSE: [ FTS3Files] })
+
+  """
+
+  _log = gLogger.getSubLogger("selectUniqueRandomSource")
+
+  # destGroup will contain for each target SE a dict { source : [list of FTS3Files] }
+  groupBySource = {}
+
+  # For all files, check which possible sources they have
+  res = _checkSourceReplicas(ftsFiles)
+  if not res['OK']:
+    return res
+
+  filteredReplicas = res['Value']
+
+  for ftsFile in ftsFiles:
+
+    if ftsFile.lfn in filteredReplicas['Failed']:
```

Comment:
If the file no longer exists, can we set the ftsFile.status to Defunct so the job gets cleaned out of the system?

Reply:
A non existing file will be filtered already at the previous step, when grouping by target. I will check what we do when this happens. However in principle it should not happen, since it would already be filtered out when going from RMS to FTS

Reply:
Hum, no such filtering at the moment WITHIN FTS, but as I said, it should never enter FTS then. Would you prefer me to add more checks for that ?

Reply:
I had it happen that some file was no longer existing and the Operation/Job got stuck in a cycle.
It could be that that was caused by the debugging activities.

Reply:
The problem occurred like this:
Created a production to move files from A to B
Some files always failed, because they were not correctly stored on A, the file size was 0. 
I deleted the file. Thus it was already in the FTS system but no longer existing.